### DISCUSSION TITLE
Trying to understand double buffer

### DISCUSSION THREAD

**Discussion by Westworld** [LAutour/ESP32-HUB75-MatrixPanel-I2S-DMA-icn2053#7] [Q&A] [answered]
In the discussion "128x128 Panel, FM6363, ICND2018, HUB75 #6" you write that double buffer can be disabled.

I know that from the "original" ESP32-HUB75-MatrixPanel-I2S-DMA library, where using DrawPixelRGB888 just (sooner or later) changes the pixel on the panel, no flipDMABuffer is required.

For this library I always thought double buffer is mandatory. Your answer in the thread named above gives the feeling that it can be disabled? 
Is there a special trick?
I used your rainbow example and set only double buffer (and also both options) to off.
Then removed each dma_display->flipDMABuffer()

The panel starts full white (while the serial port shows that step by step the colors are called), staying white. No drawing occurs.

Any hint what is missing would be great.

**Comment by LAutour** [comment, marked as answer]
In the ESP32-HUB75-MatrixPanel-I2S-DMA, the same buffer is displayed in which we draw (if there are two, then we draw in one, the second is displayed.)

For panels with smart drivers (with memory) there is no way to combine the buffer for drawing and for output (16 bits for each color channel - just not enough memory). Therefore, a separate full buffer for drawing and a separate reduced dma buffer for output are used. 
The flipDMABuffer() command (the name has not been changed for compatibility) starts the transfer of parts of data from the full drawing buffer to the output dma buffer (with recoding). 

The double buffer duplicates the drawing buffer and allows you to draw in one while the data from the other is displayed on the screen. The double dma buffer allows you to prepare recoded data for output while another dma buffer is being output. Both double buffer and double dma buffer are used only to speed up the image change at the expense of more memory (and do not always give a noticeable effect).

**Reply by Westworld** [reply]
many thanks for the fast answer!

**Reply by LAutour** [reply]
When the double drawing buffer is disabled, in addition to flipDMABuffer(), you can use sendFrame(bool waitSend, bool autoVsync), where waitSend - whether to wait for the end of the output in the panel, autoVsync - whether to immediately show the sent image on the panels (the panels themselves have a double image buffer). If autoVsync = false, then you will have to send the sendVsync () command (toggle the befer in the panels)
This is convenient if you need to make a clock with synchronous output: send an image to the panel in advance (relatively long), and then output it at the right time (relatively fast).
